core/templates/domain/exploration/InteractionCustomizationArgObjectFactory.ts

```diff
@@ -0,0 +1,135 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating an interaction customization argument.
+ */
+
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+
+import { Schema } from 'services/schema-default-value.service';
+import {
+  SubtitledHtml, SubtitledHtmlObjectFactory, SubtitledHtmlBackendDict
+} from 'domain/exploration/SubtitledHtmlObjectFactory';
+import {
+  SubtitledUnicode, SubtitledUnicodeObjectFactory, SubtitledUnicodeBackendDict
+} from 'domain/exploration/SubtitledUnicodeObjectFactory';
+import {
+  InteractionCustomizationArgsBackendDictValue,
+  InteractionCustomizationArgsValue
+} from 'interactions/customization-args-defs';
+import { cloneDeep } from 'lodash';
+import { SchemaConstants } from
+  'components/forms/schema-based-editors/schema-constants';
+
+export interface InteractionCustomizationArgBackendDict {
+  value: InteractionCustomizationArgsBackendDictValue
+}
+
+/**
+ * InteractionCustomizationArg represents a single customization argument.
+ * Customization arguments for an interaction are stored in a dictionary from
+ * customization argument name to a InteractionCustomizationArg object.
+ */
+export class InteractionCustomizationArg {
+  constructor(public value: InteractionCustomizationArgsValue) {}
+
+  toBackendDict(): InteractionCustomizationArgBackendDict {
+    const traverseSchemaAndConvertSubtitledToDicts = (
+        value: Array<Object> | Object
+    ): InteractionCustomizationArgsBackendDictValue => {
+      let result: InteractionCustomizationArgsBackendDictValue;
+
+      if (value instanceof SubtitledUnicode || value instanceof SubtitledHtml) {
+        result = value.toBackendDict();
+      } else if (value instanceof Array) {
+        result = value.map(element =>
+          traverseSchemaAndConvertSubtitledToDicts(element));
+      } else if (value instanceof Object) {
+        result = {};
+        Object.keys(value).forEach(key => {
+          result[key] = traverseSchemaAndConvertSubtitledToDicts(value[key]);
+        });
+      }
+
+      return result || value;
+    };
+
+    return {
+      value: traverseSchemaAndConvertSubtitledToDicts(this.value)
+    };
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class InteractionCustomizationArgObjectFactory {
+  constructor(
+    private subtitledHtmlObjectFactory: SubtitledHtmlObjectFactory,
+    private subtitledUnicodeObjectFactory: SubtitledUnicodeObjectFactory
+  ) {}
+
+  createFromBackendDict(
+      customizationArgBackendDict:
+        InteractionCustomizationArgBackendDict,
+      customizationArgSchema: Schema
```

Comment:
I still think this function can be refactored to use discriminated unions. i.e. `customizationArgBackendDict` can take values according to the `customizationArgSchema`. Currently you've just typecasted this variable everywhere. 

Mb if you can brief me about what this function does and where is this used, I would be able to help you with this.

Reply:
Hmmm, I tried refactoring it. But seems like discriminated unions don't work when the discriminating factor is nested. In this case, it would be `arg.schema.type`. I think to use discriminated unions here we would need to pass on the `schema_type` separately and there would be problems in the `dict` schema if we do so. Maybe @brianrodri can advice here.

Reply:
Hmm, after doing some digging, it seems like TypeScript generics aren't powerful enough to create the perfect "schema" type we need here (unlike C++, which is powerful enough to do it through templates), and it won't be fixed in the forseeable future: https://github.com/microsoft/TypeScript/issues/5453#issuecomment-646746881.

We'll have to settle for hard-coded types instead. Given that the number of interactions is relatively small, it seems reasonable for the short-term.

In the example you've provided, we can expose the schema as simply:

```typescript
interface TextInputCustomizationArgSchema {
  type: 'list';
  items: {
    type: 'dict';
    properties: [{
      name: 'content';
      schema: {
        type: 'custom';
        obj_type: 'SubtitledUnicode';
      }
    }, {
      name: 'show';
      schema: {
        type: 'bool';
      };
    }];
  };
};
```

Going with this strategy, the values of TextInputs are **not** flexible. The best we can do is mandate that input schemas are exactly the same as this one. For different versions, we can provide the same schema with slight differences:

```typescript
interface TextInputCustomizationArgSchemaV1 {
  version: 1;
  // ...
}

interface TextInputCustomizationArgSchemaV2 {
  version: 2;
  // ...
}

type TextInputCustomizationArgSchema = (
  TextInputCustomizationArgSchemaV1 |
  TextInputCustomizationArgSchemaV2);
```

This is the closest we can get to perfect type safety :(

Does this unblock you?

Reply:
Sure, that makes sense. How do we plan on combining the schemas for all customization argument schemas? Will they just be unioned together? @brianrodri

Reply:
Yup, that's right. Just please make sure that the interaction id is at the top of each schema so we can use it as a discriminant.

Reply:
Hi all, from my discussion with @brianrodri  it seems we won't need to recursively traverse the schema anymore, so won't need Schema types here. Done